### PR TITLE
Add configuration option for adding or prepending repositories

### LINKED ISSUE #95: Merging VCS repository override of published package fails
First reported by @ghprod in https://github.com/wikimedia/composer-merge-plugin/issues/71#issuecomment-154435583

**composer.json:**
```
{
    "require": {
        "wikimedia/composer-merge-plugin": "dev-master"
    },
    "config": {
        "preferred-install": "dist"
    },
    "extra": {
        "merge-plugin": {
            "include": "composer.local.json"
        }
    }
}
```

**composer.local.json:**
```
{
    "repositories": [
        {
            "type": "vcs",
            "url": "https://github.com/ghprod/laravel-menu"
        }
    ],
    "require":{
        "lavary/laravel-menu": "dev-master#956e50f"
    }
}
```

**Result:**
```
$ composer -vvv update
[...snip...]
Loading plugin Wikimedia\Composer\MergePlugin
  [merge-plugin] Loading composer.local.json...
  [merge-plugin] Adding vcs repository github.com/ghprod/laravel-menu
  [merge-plugin] Merging lavary/laravel-menu
Loading composer repositories with package information
[...snip...]
Downloading https://api.github.com/repos/ghprod/laravel-menu
Downloading https://api.github.com/repos/ghprod/laravel-menu/contents/composer.json?ref=master
Downloading https://api.github.com/repos/ghprod/laravel-menu/commits/master
Downloading https://api.github.com/repos/ghprod/laravel-menu/tags?per_page=100
Downloading https://api.github.com/repos/ghprod/laravel-menu/git/refs/heads?per_page=100
[...snip...]
Reading composer.json of lavary/laravel-menu (master)
Downloading https://api.github.com/repos/ghprod/laravel-menu/contents/composer.json?ref=956e50fbf0d99b2804539069b012ad52b4851dd2
Downloading https://api.github.com/repos/ghprod/laravel-menu/commits/956e50fbf0d99b2804539069b012ad52b4851dd2
[...snip...]
Downloading http://packagist.org/p/lavary/laravel-menu%24cb8432bd74b707af0300352a1d49742389810f45e5646550e5dca3a823fe4c55.json
[...snip...]
Analyzed 1426 packages to resolve dependencies
Analyzed 1200 rules to resolve dependencies
[...snip...]
  - Installing lavary/laravel-menu (dev-master bebce99)
Downloading https://api.github.com/repos/lavary/laravel-menu/zipball/bebce996ef08cf911da0139c30a2da03a2768326
```

The installed revision (bebce99) is the head of the dev branch of the project as listed on Packagist. The alternate VCS repository was not used.

If the require for "lavary/laravel-menu" is moved to composer.json, results are slightly different but still not correct. In that case the logs report:
```
  - Installing lavary/laravel-menu (dev-master 956e50f)
Downloading https://api.github.com/repos/lavary/laravel-menu/zipball/956e50f
```

This download is still from the wrong repository in this case and the files fetched from GitHub are not not actually for the desired hash.

If the "repositories" are moved to composer.json instead, the new output is:
```
  - Installing lavary/laravel-menu (dev-master 956e50f)
Downloading https://api.github.com/repos/ghprod/laravel-menu/zipball/956e50fbf0d99b2804539069b012ad52b4851dd2
```

This is the desired result for all three tests.

I think the issue here is the order that the repositories are added to `Composer\Repository\RepositoryManager`. It may take an upstream patch to fix this properly.

### PR DESCRIPTION
**Pull request by Furgas**
I think this should fix #95.

### CODE REVIEW

**Comment by LionsAd** [comment]
Looks great to me!

File: src/Merge/ExtraPackage.php [hunk 1]
```diff
@@ -192,20 +193,35 @@ protected function addRepositories(RootPackageInterface $root)
             if (!isset($repoJson['type'])) {
                 continue;
             }
-            $this->logger->info("Adding {$repoJson['type']} repository");
+            if ($prepend) {
+                $this->logger->info("Adding {$repoJson['type']} repository");
+            } else {
+                $this->logger->info("Prepending {$repoJson['type']} repository");
```
Comment (bd808) [inline src/Merge/ExtraPackage.php]: These log messages are swapped.

**Comment by bd808** [comment]
In local testing I can see the proper functions (append/prepend) being called, but I'm not having success getting the reproduction case from issue #95 to actually resolve as hoped.

The feature flag that you have added is not documented, but I think if we can get it to work for the test case then we actually don't need to have a feature flag. The desired behavior would be for repos added by the included/required config files to have a higher priority than the repos added in the root composer.json and certainly higher than the default packagist repo.

**Comment by Furgas** [comment]
It seems to work for me.

**composer.json**
```
{
    "repositories": [
        {
            "type": "git",
            "url": "https://github.com/Furgas/composer-merge-plugin"
        }
    ],
    "require": {
        "wikimedia/composer-merge-plugin": "dev-add-or-prepend-repo"
    },
    "config": {
        "preferred-install": "dist"
    },
    "extra": {
        "merge-plugin": {
            "include": "composer.local.json",
            "prepend-repositories": true
        }
    }
}

```

**composer.local.json**
```
{
    "repositories": [
        {
            "type": "vcs",
            "url": "https://github.com/ghprod/laravel-menu"
        }
    ],
    "require": {
        "lavary/laravel-menu": "dev-master#956e50f"
    }
}
```

Inside **composer.lock** after `composer install`:
```
{
    "name": "lavary/laravel-menu",
    "version": "dev-master",
    "source": {
        "type": "git",
        "url": "https://github.com/ghprod/laravel-menu.git",
        "reference": "956e50f"
    },
    "dist": {
        "type": "zip",
        "url": "https://api.github.com/repos/ghprod/laravel-menu/zipball/956e50f",
        "reference": "956e50f",
        "shasum": ""
    },
    "require": {
        "illuminate/support": ">=5.0",
        "illuminate/view": ">=5.0",
        "php": ">=5.3.0"
    },
    "type": "library",
    "autoload": {
        "psr-0": {
            "Lavary\\Menu\\": "src/"
        }
    },
    "license": [
        "MIT"
    ],
    "authors": [
        {
            "name": "Lavary",
            "email": "mrl.8081@gmail.com"
        }
    ],
    "description": "A quick way to create menus in Laravel 5",
    "homepage": "https://github.com/lavary/laravel-menu",
    "keywords": [
        "laravel"
    ],
    "support": {
        "source": "https://github.com/ghprod/laravel-menu/tree/master"
    },
    "time": "2015-11-14 15:14:37"
}
```

**Comment by bd808** [comment]
> It seems to work for me.

In retesting it seems to work for me as well. I think when I tested previously I must have missed changing the wikimedia/composer-merge-plugin requirement to point to the correct branch

**Comment by alexweissman** [comment]
Any plans for a release with this update any time soon?  I'm running into the same problem from #95 with trying to override Packagist with a local version in a merged composer file.

**Pull request merged**